### PR TITLE
generalize experiment runner for multiple model classes

### PR DESCRIPTION
**Pull request by BoltzmannBrain**

### CODE REVIEW

**Comment by BoltzmannBrain** [comment]
@scottpurdy please review.

File: fluent/encoders/cio_encoder.py [hunk 1]
```diff
@@ -59,33 +59,20 @@ def encode(self, text):
...
+                                      encoding["fingerprint"]["positions"].
     """
-    if isinstance(text, str):
-      text = self.client.tokenize(text)
-
     try:
-      encoding = self.client.getBitmap(text)
-    except ValueError:
       encoding = self.client.getTextBitmap(text)
+    except Exception:
```
Comment (scottpurdy) [inline fluent/encoders/cio_encoder.py]: Can you make this more specific?

File: fluent/encoders/cio_encoder.py [hunk 2]
```diff
@@ -111,6 +97,24 @@ def decode(self, encoding, numTerms=None):
...
+    @param text             (str)             A non-tokenized sample of text.
+    @return encoding        (dict)            Fingerprint from cortipy client.
+                                              An empty dictionary of the text
+                                              could not be encoded.
+    """
+    tokens = list(itertools.chain.from_iterable(
+      [t.split(',') for t in self.client.tokenize(text)]))
+    try:
+      encoding = min([self.client.getBitmap(t) for t in tokens],
+        key=lambda x: x["df"])
```
Comment (scottpurdy) [inline fluent/encoders/cio_encoder.py]: indentation

File: fluent/encoders/cio_encoder.py [hunk 3]
```diff
@@ -111,6 +97,24 @@ def decode(self, encoding, numTerms=None):
...
+    except Exception:
```
Comment (scottpurdy) [inline fluent/encoders/cio_encoder.py]: Can you make this more specific?

File: fluent/experiments/baseline_runner.py [hunk 4]
```diff
@@ -161,7 +164,13 @@ def run(args):
       model = pkl.load(f)
     print "Model loaded from \'{0}\'.".format(modelPath)
   else:
-    model = ClassificationModelRandomSDR(verbosity=args.verbosity)
+    try:
+      module = __import__(args.modelModuleName, {}, {}, args.modelName)
```
Comment (scottpurdy) [inline fluent/experiments/baseline_runner.py]: Why do you have to use `__import__`?

Comment (BoltzmannBrain) [inline fluent/experiments/baseline_runner.py]: To dynamically import the model that is specified as an argument to the experiment runner. This way the same experiment runner can be used for `ClassificationModelRandomSDR` and `ClassificationModelFingerprint`, and any future `ClassificationModel` subclasses.

File: fluent/models/classify_fingerprint.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+  def trainModel(self, sample, label):
+    """
+    Train the classifier on the input sample and label.
+
+    @param sample     (numpy.array)     Bitmap encoding of the sample.
+    @param label      (int)             Reference index for the classification
+                                        of this sample.
+    """
+    if sample.any():
+      _ = self.classifier.learn(sample, label, isSparse=self.n)
```
Comment (scottpurdy) [inline fluent/models/classify_fingerprint.py]: No need to do anything with the return value. If you don't store it in a variable it just gets cleaned up

File: fluent/models/classify_fingerprint.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+    @param sample     (numpy.array)     Bitmap encoding of the sample.
+    @return           (list)            The n most-frequent classifications
+                                        for the data samples; for more, see the
+                                        KNNClassifier.infer() documentation.
+                                        Values are int or None.
+    Note: to return multiple winner classifications, modify the return statement
+    accordingly.
+    """
+    tokenLabels = []
```
Comment (scottpurdy) [inline fluent/models/classify_fingerprint.py]: this isn't used

File: fluent/models/classify_random_sdr.py [hunk 7]
```diff
@@ -77,10 +83,9 @@ def trainModel(self, sample, label):
     """
     # This experiment classifies individual tokens w/in each sample. Train the
     # kNN classifier on each token.
-    p = 0
     for bitmap in sample:
       if bitmap == []: continue
-      p = self.classifier.learn(bitmap, label, isSparse=self.n)
+      _ = self.classifier.learn(bitmap, label, isSparse=self.n)
```
Comment (scottpurdy) [inline fluent/models/classify_random_sdr.py]: again, if you aren't using the return value no need to store it

**Comment by scottpurdy** [comment]
We should really stick to using the NuPIC random number generator. If there is functionality you want, like seeding it with a string, then it should be pretty simple to add.

**Pull request merged**